### ISSUE TITLE
Move away from null checks and .SingleOrDefault checks in controllers/handlers and instead rely on validating this information up front

### ISSUE THREAD

**Issue by mgmccarthy** [HTBox/allReady#1257]
@BillWagner @MisterJames @tonysurma @sanilpaul @stevejgordon @VishalMadhvani 

This is an overall pattern I've noticed in allReady, where every time we go to query something, we're always allowing the return of nulls and then building subsequent conditionals to handle those nulls in consuming code. This additional overhead of checking for nulls carries into our unit tests, and makes our classes do more work than they have to.

For an example, let's take the following files:
- AllReady.Areas.Admin.Controllers.CampaignController
- CampaignSummaryQueryHandlerAsync
- associated unit tests

When CampaignController sends CampaignSummaryQueryHandlerAsync, the user is requesting an edit action method which will return to them a screen where they can change information about the Campaign:

Here is the controller code:
```
public async Task<IActionResult> Edit(int id)
{
    var viewModel = await _mediator.SendAsync(new CampaignSummaryQueryAsync { CampaignId = id });
    if (viewModel == null)
    {
        return NotFound();
    }
...
```

Here is the handler code:
```
public async Task<CampaignSummaryViewModel> Handle(CampaignSummaryQueryAsync message)
{
    CampaignSummaryViewModel result = null;

    var campaign = await _context.Campaigns
        .AsNoTracking()
        .Include(ci => ci.CampaignImpact)
        .Include(mt => mt.ManagingOrganization)
        .Include(l => l.Location)
        .Include(c => c.CampaignContacts).ThenInclude(tc => tc.Contact)
        .SingleOrDefaultAsync(c => c.Id == message.CampaignId)
        .ConfigureAwait(false);
...
```

When, under any circumstance, would we not be able to find the Campaign we want to edit by id? How is this id provided? It's provided by a previous query/action method on the controller. 

The way that campaign got into the system is we had a page that allowed a user to create a campaign, and validated everything up front before the creation was allowed. Basically, the UI enforces this. The Campaign should be there when we go to list all Campaigns, and it should also be there when a user picks a Campaign from this list to edit.

Sure, you could cobble together a URL like this:
`\Admin\Campaign\Edit\5`

and paste it into the address bar of your browser and hit enter, but if someone is doing that, they're most likely up to no good. So why are we optimizing the user experience to be friendly (we're going to return a nice `NotFound()` result) for someone is most likely up to no good?

The answer is we shouldn't. We should be optimizing for the happy path.

That being said, here is an example of additional unit tests we've been adding to the system to test these null returns:
```
public class CampaignAdminControllerTests
{
    [Fact]
    public async Task EditGetReturnsHttpNotFoundResultWhenViewModelIsNull()
    {
        CampaignController sut;
        MockMediatorCampaignSummaryQuery(out sut);
        Assert.IsType<NotFoundResult>(await sut.Edit(It.IsAny<int>()));
    }
}
```

I'd like to see a fundamental shift if this practice to optimize for the happy path. The chances of a user going to edit a Campaign that they just loaded 3 seconds ago from and Index action method being missing is extremely low. The only way I could see that happening is if some other user deleted the same campaign while they were in middle of their workflow, and again, I see this being very low.

Thoughts?

**Comment by stevejgordon** [comment]
@mgmccarthy My initial opinion is that I prefer the current approach. Yes it adds a bit of code, but ultimately we are coding for all possible cases and ensuring we return a valid and suitable response. I agree the scenarios which could lead to controllers being hit with invalid id's should be very limited but it's the web and it's still a possibility.

In your example, what would we return instead of null from the handler in the case that an invalid id was provided (either intentionally or due to some unforeseen issue/bug elsewhere)? What does the controller do as it's expecting something back from the handler. If for any reason we haven't got something to give it, what will the controller return?

I've not found it particularly painful to code with the current style and it is then explicit in what's expected to happen.

**Comment by mgmccarthy** [comment]
@stevejgordon 

"what would we return instead of null from the handler in the case that an invalid id was provided (either intentionally or due to some unforeseen issue/bug elsewhere)?"
we allow the handler to throw an exception

"What does the controller do as it's expecting something back from the handler. If for any reason we haven't got something to give it, what will the controller return?"
we allow the controller to throw an exception when it goes to use any of the properties on the returned object

In both cases, the user can be redirected to whatever allReady's error page is. MediatR, although a CQRS framework, executes its pipeline synchronously and if one of the handlers in a list of messages/handlers fails, everything rolls back. Since all messages sent can be traced back to a message sent from a controller, that exception will travel back to the controller.

We should be able to use something like [Elmah ](https://elmah.io/?gclid=CjwKEAjwgo6_BRC32q6_5s2R-R8SJAB7hTG-Y8N2OUn_kndfpfMVx6oGJTyVDf_Vkihpg-4rJ51QuhoCN7vw_wcB) to capture and allow us look up the causes of errors and troubleshoot accordingly (unless there is already something in place for the project. I know we have `ILogger<T>` injected into parts of the system, I'm just not too sure where that gets logged, and how that log is accessed).

The point here is if an exception is thrown, it's most likely going to be thrown for something that is out of the ordinary (aka, not regular usage of the site)...

also, if we check for null on everything, **how do we differentiate between a null that is expected as part of the business case/use case for a given operation, and a null as the result of "something going wrong"**. 

This makes existing code doing null checks a little harder to understand. Are we checking for null because we allowed a null to be written to the db as part of a write operation against whatever we're querying, or are we checking for null "just in case it's null".

Checking for null should be done when the use case expects a null as part of its functionality, not just to cover our behinds.

Obviously, this is all IMHO, but I figured I'd raise the issue for discussion.

**Comment by sanilpaul** [comment]
I understand the more cautious approach of having guard clauses. if we are following CQRS approach, there shouldn't be any logical reason for a command to fail. So all the elements in the command should be validated upfront. In this specific instance, it looks like Campaign gets validated upfront. No one is entering CampaignId in a free flowing text and its not a good idea to optimize for malicious user. So I would prefer in the handler to be `.Single`. So in case, someone tries to edit a campaign that doesn't exist, it would throw exception as it truly is an exceptional case. The Pros are less code/test in the handler & the intent in the handler will be clear.
		 
It looks like `if (viewModel == null)` check in the controller is due to the `.SingleOrDefault` in the handler. now the side effect of null check trickles down. we have to write even more tests. We are returning `404 - NotFound()`. Instead, I think what Mike's proposing is to lets use what we get for free from the framework aka error page for that very exceptional case. When I started writing tests for another handler, it truly baffled me that the Id passed through via command may or may not exist. I think eliminating those extra checks in places where its not necessary would clean up the code base as well as make the intent clear to some one who reads the code.

**Comment by VishalMadhvani** [comment]
For me the general rule of thumb for web development is to not trust user input.
In this case, the controller action methods are at the edge and hence are responsible for validating user input.

Within the system, there isn't a need for defensive programming.. exceptions can be caught by the framework, logged and then handled by showing an error page.
Whether this is done by exceptions or by passing nulls in my opinion should be left to the individual developer.

**Comment by BillWagner** [comment]
In general, I'd prefer the simpler solution of throwing exceptions when there are errors in user data or code.

However, I have a caveat and a question:  If we throw exceptions, instead of our current plan, do we give a malicious user information that would help them attack the site? That would be the reason to catch these errors and redirect to an innocuous error page.

**Comment by mgmccarthy** [comment]
For handling exceptions (aka, not Dev or Staging), it looks like we already have code to do that in Startup.cs:
![image](https://cloud.githubusercontent.com/assets/2428993/18887243/4f3bd9f4-84c1-11e6-9bbc-c615c10347fc.png)

This is what the HomeController/Error page looks like:
![image](https://cloud.githubusercontent.com/assets/2428993/18887294/7ac6a900-84c1-11e6-94f9-c57a908e6d2a.png)

so based on what environment you're in, you'll see the specific exception details, or the generic error page. As long as the environment is configured properly, a malicious user should be able to see the specifics of any exceptions.

Granted, I still have to check out if throwing an exception will land the user on that generic error page for the given environment, but I can look into that using the following asp.net docs:
https://docs.asp.net/en/latest/fundamentals/error-handling.html#configuring-an-exception-handling-page.

**Comment by tonysurma** [comment]
@stevejgordon @mgmccarthy @BillWagner @MisterJames is there a consensus on what the next steps/issues would be here?

**Comment by stevejgordon** [comment]
@tonysurma From my perspective, @mgmccarthy has made some very good points and he's winning me around. I've been working a lot with APIs recently where it's more important to return correct status codes etc so I was a bit led by that approach.

Personally I still like checking things and avoiding throwing exceptions but I can see the merit in simplifying the logic which in turn simplifies the tests. Perhaps we can discuss on the standup?

**Comment by mgmccarthy** [comment]
@tonysurma, I'm still in favor of getting rid of what I consider a lot of redundant, and often confusing (especially for new contributors) null checks.

However, I would like to hear @MisterJames @BillWagner and @dpaquette weigh in on the issue.

Refactoring this code would result in pretty much code deletion and less unit tests. We would start with a handler, change it from SingleOrDefault to Single (this is one example), then work our way backwards to the controller that sends the message the handler handles.

**Comment by dpaquette** [comment]
@mgmccarthy we did talk about this on one of the weekly standups. I think @MisterJames was planning to write up summary of our thoughts.

**Comment by mgmccarthy** [comment]
The caveat with doing this is we need some type of clearly defined requirement that tells us  which pages can be viewed by an unauthorized user... @stevejgordon just had to fix #1302 b/c my assumption when working in that area is that an unauthorized user would not be able to view an Event.

**Comment by mgmccarthy** [comment]
@MisterJames @BillWagner and @dpaquette, a quick amendment to my previous comment... we don't need some type of clearly identified requirement to tell contributors which pages can be viewed by an unauthorized user... it's easy to figure this out with usage of the [Authorize] and [AllowAnonymous] attributes.

**Comment by hodorswit** [comment]
I've had some success in other projects dealing with this issue by leveraging ActionFilters and ExceptionFilters. The ActionFilter would return a NotFoundResult if a controller action returns a null IActionResult, and the ExceptionFilter(s) would convert domain-specific exceptions into HTTP status codes. This got rid of most of the conditional checks in the controller. 

The drawback to this approach was that it tended to confuse new developers. It is not everyone's first intuition to look for globally registered filters that may be modifying their request/response, so it can feel like "magic". 

I'm new here, by the way. Been reading through issues for the last couple days and trying to orient myself. Looking forward to getting more involved!

**Comment by mgmccarthy** [comment]
@hodorswit, thanks for the input. 

cc @stevejgordon @tonysurma @shahiddev @BillWagner @MisterJames 

This is something we definitely can deal with using ActionFilters and ExceptionFilters, but like you said, if might not be apparent to new developers that this "magic" is happening in the filters, event though most .net developers are familiar with the framework-supplied filters that we see decorated controller classes and action methods.

That being said, one of the goals of the project (partially b/c it's open source), is to optimize for the new contributor "F5" experience (where one can pull, then stand up the system and start running it immediately) as well as keep the cognitive load as low as possible to enable people that want to contribute at a very shallow level (for example, writing unit tests), or go deep (for example, invoking web-based API's, figuring out the contract, wrapping it in an abstraction, etc...)...

The more and more I think about this Issue, I know believe it is tied to Issue #873, which is our ongoing discussion of what logging framework to use in production.

It seems that an Azure-based logging might when out over my proposal of ELMAH.IO, b/c it has native support for the application insights and telemetry that .NET Core provides.

But either way, regardless of the logging platform (ELMAH.IO/Azure logging), that logging framework automatically should be able to take exceptions that are not handled, log them for you, allow you define a compensating action (in production, re-direct to generic error page). 

On the back-end, the logging framework should have built in dashboards as well as hooks for email/sms notifications to go out letting people know something went wrong, so in a way. 

So, in summary, a mature logging framework already accomplishes what Action and Exception filters would by hand.

That being said, that does not rule out us taking the Action/Exception filter approach, as this approach still has it's merits. We could even inject and ILogger<T> instance into these filters to be more explicit about exceptions we'd like to capture, what "level" of errors that are (info, warn, etc...) and how they should be logged.
